### ISSUE TITLE
recode=TRUE 

### ISSUE THREAD

**Issue by arilamstein** [walkerke/tidycensus#283]
I am a bit surprised by how the `recode` parameter works in `get_pums`.

Some variables (such as "COW" (Class of Worker)) have coded variables, which you can't possibly understand without help:

```
> df_cow = get_pums("COW", "wyoming", 2018, "acs1")
Getting data from the 2018 1-year ACS Public Use Microdata Sample
Using FIPS code '56' for state 'Wyoming'
Downloading: 240 kB     > df_cow
# A tibble: 5,740 x 6
   SERIALNO      SPORDER  WGTP PWGTP ST    COW  
   <chr>           <dbl> <dbl> <dbl> <chr> <chr>
 1 2018GQ0004385       1     0    38 56    4    
 2 2018GQ0006396       1     0    11 56    b    
 3 2018GQ0010505       1     0    47 56    1    
 4 2018GQ0010943       1     0     6 56    1    
 5 2018GQ0017389       1     0    44 56    1    
 6 2018GQ0021231       1     0   190 56    4    
 7 2018GQ0021234       1     0     6 56    1    
 8 2018GQ0025647       1     0   296 56    b    
 9 2018GQ0026496       1     0    94 56    1    
10 2018GQ0027272       1     0   328 56    4    
```

I love that there is an option `recode=TRUE` that helps with this, but it behaves slightly differently than I expected:
```
> df_cow2 = get_pums("COW", "wyoming", 2018, "acs1", recode=TRUE)
Getting data from the 2018 1-year ACS Public Use Microdata Sample
Using FIPS code '56' for state 'Wyoming'
Downloading: 240 kB     
> df_cow2
# A tibble: 5,740 x 8
   SERIALNO    SPORDER  WGTP PWGTP ST    COW   ST_label  COW_label                                                   
   <chr>         <dbl> <dbl> <dbl> <chr> <chr> <chr>     <chr>                                                       
 1 2018GQ0004…       1     0    38 56    4     Wyoming/… State government employee                                   
 2 2018GQ0006…       1     0    11 56    b     Wyoming/… N/A (less than 16 years old/NILF who last worked more than …
 3 2018GQ0010…       1     0    47 56    1     Wyoming/… Employee of a private for-profit company or business, or of…
 4 2018GQ0010…       1     0     6 56    1     Wyoming/… Employee of a private for-profit company or business, or of…
 5 2018GQ0017…       1     0    44 56    1     Wyoming/… Employee of a private for-profit company or business, or of…
 6 2018GQ0021…       1     0   190 56    4     Wyoming/… State government employee                                   
 7 2018GQ0021…       1     0     6 56    1     Wyoming/… Employee of a private for-profit company or business, or of…
 8 2018GQ0025…       1     0   296 56    b     Wyoming/… N/A (less than 16 years old/NILF who last worked more than …
 9 2018GQ0026…       1     0    94 56    1     Wyoming/… Employee of a private for-profit company or business, or of…
10 2018GQ0027…       1     0   328 56    4     Wyoming/… State government employee                                   
# … with 5,730 more rows
```

My client wanted a total replacement of the value codes with the value labels. The code for this isn't complex, but as I suspect other people might want it, it might be nice if `recode` supported this in some way.

More substantially, many people on my team wanted the same treatment for the column names. (I.e. replacing the variable codes with the variable labels. I wrote a function called `rewrite_variables` that does this. 

```
> rewrite_variables(df_cow2, vars_2018) 
# A tibble: 5,740 x 8
   `SERIALNO: Hous… `SPORDER: Perso… `WGTP: Housing … `PWGTP: Person'…
   <chr>                       <dbl>            <dbl>            <dbl>
 1 2018GQ0004385                   1                0               38
 2 2018GQ0006396                   1                0               11
 3 2018GQ0010505                   1                0               47
 4 2018GQ0010943                   1                0                6
 5 2018GQ0017389                   1                0               44
 6 2018GQ0021231                   1                0              190
 7 2018GQ0021234                   1                0                6
 8 2018GQ0025647                   1                0              296
 9 2018GQ0026496                   1                0               94
10 2018GQ0027272                   1                0              328

> colnames(rewrite_variables(df_cow2, vars_2018))
[1] "SERIALNO: Housing unit/GQ person serial number" 
[2] "SPORDER: Person number"                         
[3] "WGTP: Housing Unit Weight"                      
[4] "PWGTP: Person's weight"                         
[5] "ST: State Code based on 2010 Census definitions"
[6] "COW: Class of worker"                           
[7] "ST_label"                                       
[8] "COW_label"  
```

Reading long column names in R is hard, but others on the team are using software (such as Excel and Tableau) where that is not an issue.

In short, it might be useful if `recode` became two options: `recode_variables` and `recode_values`. And it might be nice if the existing functionality (recoding values) went a step further and had the option to overwrite the value codes altogether.

**Comment by arilamstein** [comment]
Oh, and for the sake of completeness, here is the code I wrote for my own functions that doe this. Please feel free to use if it helps you:

```
# The column names are variable codes - e.g. "COW"
# We want to replace them with the variable code + the variable label
# E.g. "COW: Class of Worker"
rewrite_variables = function(df_pums_data, df_pums_vars)
{
  col_names = colnames(df_pums_data)
  for (i in 1:ncol(df_pums_data)) 
  {
    # if this column name is present in the variable data frame as a "variable code"
    col_name = colnames(df_pums_data)[i]
    if (col_name %in% df_pums_vars$var_code)
    {
      # get the "variable label"
      var_label = df_pums_vars[df_pums_vars$var_code == col_name, "var_label"]
      
      # set the column name as something that contains both code and label
      colnames(df_pums_data)[i] = paste0(col_name, ": ", var_label)
    }
  }
  df_pums_data
}

library(stringr)

# Replace all Values (e.g. "0", "1") with their Value Labels (e.g. "Male", "Female")
# To do this we exploit the fact that all encoded variables are in a column that ends in 
# "_label". E.g. Replace the column "SEX" with the contents of the column "SEX_label" 
# and wrap up by deleting the column "SEX_label"
rewrite_values = function(df)
{
  all_col_names   = colnames(df)
  label_col_names = all_col_names[str_detect(all_col_names, "_label")]
  
  for (cn in label_col_names)
  {
    # the "original name" is the one before the prefix "_label"
    original_name = str_split(cn, "_")[[1]][1]
    # override the codes with the values
    # then delete the code column
    df[,original_name] = df[,cn]
    df[,cn] = NULL
  }
  
  df
}
```

**Comment by walkerke** [comment]
Thanks for your comments, @arilamstein.  I appreciate you sharing this code as well as I think it'll be useful to other users!  I'd be open to some discussion on these matters but a couple first thoughts: 

- I'm unsure of the necessity of over-writing a value column with its labels when a user could instead request `recode = TRUE` and then drop the value column themselves if they don't want it in the dataset.  To me, this doesn't seem to rise to the level of a feature.  An alternative approach used by {ipumsr} and other microdata approaches in SPSS/Stata is to use labelling functionality in the software or a labelled class (implemented in R in the {haven} package).  We have an open PR that I need to merge that uses ordered factors in a way that I like: https://github.com/walkerke/tidycensus/pull/274.  

- Perhaps this is my ArcGIS background talking but I've never been a fan of long variable names with special characters (which we discussed of course over in https://github.com/walkerke/tidycensus/issues/275).  However, this is a personal preference and I'd be willing to entertain a PR that clarifies variable names for users as an option like you do with your code.  Perhaps the `janitor::clean_names()` approach suggested by @mfherman represents a path forward here?

**Comment by mfherman** [comment]
To follow up on @walkerke's comments, I agree that for most use cases, folks will probably want either just the coded values, or both the coded values and the labels. In @arilamstein's example, if `COW_label` overwrote the coded column, it would be more cumbersome to further subset the data using the label alone:

``` r
library(tidycensus)
library(dplyr, warn.conflicts = FALSE)

df <- get_pums("COW", "wyoming", 2018, "acs1", recode = TRUE)
#> Getting data from the 2018 1-year ACS Public Use Microdata Sample

df %>% 
  filter(COW_label == "Employee of a private for-profit company or business, or of an individual, for wages, salary, or commissions")
#> # A tibble: 2,067 x 8
#>    SERIALNO   SPORDER  WGTP PWGTP ST    COW   ST_label COW_label                
#>    <chr>        <dbl> <dbl> <dbl> <chr> <chr> <chr>    <chr>                    
#>  1 2018GQ000~       1     0    45 56    1     Wyoming~ Employee of a private fo~
#>  2 2018GQ000~       1     0    55 56    1     Wyoming~ Employee of a private fo~
#>  3 2018GQ001~       1     0    12 56    1     Wyoming~ Employee of a private fo~
#>  4 2018GQ002~       1     0    11 56    1     Wyoming~ Employee of a private fo~
#>  5 2018GQ003~       1     0    11 56    1     Wyoming~ Employee of a private fo~
#>  6 2018GQ003~       1     0    75 56    1     Wyoming~ Employee of a private fo~
#>  7 2018GQ003~       1     0   129 56    1     Wyoming~ Employee of a private fo~
#>  8 2018GQ003~       1     0    65 56    1     Wyoming~ Employee of a private fo~
#>  9 2018GQ004~       1     0   202 56    1     Wyoming~ Employee of a private fo~
#> 10 2018GQ004~       1     0    62 56    1     Wyoming~ Employee of a private fo~
#> # ... with 2,057 more rows

df %>% 
  filter(COW == 1)
#> # A tibble: 2,067 x 8
#>    SERIALNO   SPORDER  WGTP PWGTP ST    COW   ST_label COW_label                
#>    <chr>        <dbl> <dbl> <dbl> <chr> <chr> <chr>    <chr>                    
#>  1 2018GQ000~       1     0    45 56    1     Wyoming~ Employee of a private fo~
#>  2 2018GQ000~       1     0    55 56    1     Wyoming~ Employee of a private fo~
#>  3 2018GQ001~       1     0    12 56    1     Wyoming~ Employee of a private fo~
#>  4 2018GQ002~       1     0    11 56    1     Wyoming~ Employee of a private fo~
#>  5 2018GQ003~       1     0    11 56    1     Wyoming~ Employee of a private fo~
#>  6 2018GQ003~       1     0    75 56    1     Wyoming~ Employee of a private fo~
#>  7 2018GQ003~       1     0   129 56    1     Wyoming~ Employee of a private fo~
#>  8 2018GQ003~       1     0    65 56    1     Wyoming~ Employee of a private fo~
#>  9 2018GQ004~       1     0   202 56    1     Wyoming~ Employee of a private fo~
#> 10 2018GQ004~       1     0    62 56    1     Wyoming~ Employee of a private fo~
#> # ... with 2,057 more rows
```

As for the variable names, I'm not sure how frequently users would need this functionality, and I'd be a little wary about including an option to create non-syntactic R variable names.

**Comment by arilamstein** [comment]
That makes sense. My use case (basically, using R to ETL PUMS data which is
then viewed in Tableau and Excel by people with no prior experience with
PUMS) might be an outlier for the package.

Feel free to close the issue. At least the code is here in case anyone else
wants to use it.

**Comment by mfherman** [comment]
Thank again for all the feedback, @arilamstein. Since the `get_pums()` feature is new, it is super helpful to have folks kick the tires!
